File: packages/sdk-ts/src/resources/calls.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { Call, CallDetail, CallListQuery, CreateCall, Transcript } from '../types.js';

const enc = encodeURIComponent;

/** Calls: placing, watching and reading them. */
export class Calls {
    constructor(private readonly http: Http) {}

    /**
     * Dial a real number over your SIP trunk and connect an agent. Spends
     * credit. An `Idempotency-Key` is generated unless `opts.idempotencyKey`
     * is given; retrying with the same key returns the original call rather
     * than dialling twice. Refused with `insufficient_credits` when the
     * balance cannot cover it.
     */
    create(body: CreateCall, opts?: RequestOptions): Promise<Call> {
        return this.http.request<Call>('POST', '/v1/calls', { ...opts, body, idempotent: true });
    }

    /** Summary records (no transcript). Filter by status, direction, agent, campaign, engine, number or time window. */
    list(query: CallListQuery = {}, opts?: RequestOptions): Promise<Page<Call>> {
        return this.http.list<Call>('/v1/calls', query, opts);
    }

    /** The full record: transcript, recording, transfers, provider errors, and `failure_code`/`failure_reason` when it did not connect. */
    retrieve(id: string, opts?: RequestOptions): Promise<CallDetail> {
        return this.http.request<CallDetail>('GET', `/v1/calls/${enc(id)}`, opts);
    }

    /** Remove the log, transcript and recording. Usage rows are kept. A live call cannot be deleted. */
    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/calls/${enc(id)}`, opts);
    }

    /** End a call in progress. Billing stops when it ends. */
    hangup(id: string, opts?: RequestOptions): Promise<Call> {
        return this.http.request<Call>('POST', `/v1/calls/${enc(id)}/hangup`, opts);
    }

    transcript(id: string, opts?: RequestOptions): Promise<Transcript> {
        return this.http.request<Transcript>('GET', `/v1/calls/${enc(id)}/transcript`, opts);
    }
}

File: packages/sdk-dart/lib/src/resources/calls.dart
import '../models/call.dart';
import '../page.dart';
import '../response.dart';
import '../transport.dart';

/// Calls: placing, watching and reading them.
class Calls {
  /// Binds the resource to a transport.
  const Calls(this._transport);

  final Transport _transport;

  /// Dials a real number over your SIP trunk and connects an agent.
  ///
  /// Spends credit. An `Idempotency-Key` is generated unless
  /// [idempotencyKey] is given; retrying with the same key returns the
  /// original call rather than dialling twice.
  Future<Call> create(CreateCall body, {String? idempotencyKey}) =>
      _transport.request(
        'POST',
        '/v1/calls',
        Call.fromJson,
        body: body.toJson(),
        idempotent: true,
        idempotencyKey: idempotencyKey,
      );

  /// As [create], but keeps the request id and whether the server replayed
  /// an earlier request rather than placing a new call.
  Future<WixzelResponse<Call>> createWithResponse(CreateCall body,
          {String? idempotencyKey}) =>
      _transport.requestWithResponse(
        'POST',
        '/v1/calls',
        Call.fromJson,
        body: body.toJson(),
        idempotent: true,
        idempotencyKey: idempotencyKey,
      );

  /// Lists calls in summary form. Use [retrieve] for the transcript.
  Future<Page<Call>> list({
    int? limit,
    String? startingAfter,
    String? endingBefore,
    CallStatus? status,
    CallDirection? direction,
    String? agentId,
    String? campaignId,
    String? engine,
    DateTime? startedAfter,
    DateTime? startedBefore,
    bool? failedOnly,
    String? phoneNumber,
  }) =>
      _transport.list('/v1/calls', Call.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
        'status': status?.value,
        'direction': direction?.value,
        'agent_id': agentId,
        'campaign_id': campaignId,
        'engine': engine,
        'started_after': startedAfter,
        'started_before': startedBefore,
        'failed_only': failedOnly,
        'phone_number': phoneNumber,
      });

  /// The full record: transcript, recording, transfers, provider errors,
  /// and `failureCode`/`failureReason` when the call did not connect.
  Future<CallDetail> retrieve(String id) => _transport.request(
      'GET', '/v1/calls/${Uri.encodeComponent(id)}', CallDetail.fromJson);

  /// Removes the log, transcript and recording. Usage rows are kept. A live
  /// call cannot be deleted; hang it up first.
  Future<void> delete(String id) =>
      _transport.requestVoid('DELETE', '/v1/calls/${Uri.encodeComponent(id)}');

  /// Ends a call in progress. Billing stops when it ends.
  Future<Call> hangup(String id) => _transport.request(
      'POST', '/v1/calls/${Uri.encodeComponent(id)}/hangup', Call.fromJson);

  /// Just the transcript.
  Future<Transcript> transcript(String id) => _transport.request('GET',
      '/v1/calls/${Uri.encodeComponent(id)}/transcript', Transcript.fromJson);
}
